### PR TITLE
10401b: simplification of Relationals

### LINKED ISSUE #10401: Eq(x/(x + 1), 1) should evaluate as false

### PR DESCRIPTION
**Pull request by smichr**
ugh. The code is not simple. This is much worse than regular simplification if you try to keep a potentially invalid relationship from becoming valid, e.g. `Lt(x, x + 1)` is invalid if `x == I` so we cannot say that this relationship is automatically true. 

Some key features to keep are
1. to not reorganize the left and rhs (else an Eq that had something isolated on the lhs may no longer be that way after simplification, e.g. Eq(y, f(x)) could become Eq(y - f(x), 0))
2. to not return an expression whose character will differ from the original (e.g. not have the same singularities, etc...).

closes #10401 
closes #10404 
- [x] track down why the unrad solution changed

**References**

---
1. https://groups.google.com/forum/#!searchin/sympy/singularities$20limit/sympy/oaoAIp4DOxQ/cGZoWbf-ieUJ
2. https://github.com/sympy/sympy/wiki/Infinities-and-Singularities
3. https://github.com/sympy/sympy/issues/6091
4. https://github.com/sympy/sympy/pull/971

### CODE REVIEW

**Comment by smichr** [comment]
I think this can be made simpler. 

Some key features to keep are 

1. to not reorganize the left and rhs (else an `Eq` that had something isolated on the lhs may no longer be that way after simplification, e.g. `Eq(y, f(x))` could become `Eq(y - f(x), 0)`)
1. to not return an expression whose character will differ from the original (e.g. not have the same singularities, etc...).

**Comment by smichr** [comment]
OK, I think this is done now.

**Comment by smichr** [comment]
OK, I'm stuck. I don't know what the test passes in all but 2.7. Is there something wrong with the True, S.true test? Does _sympify(True) not return S.true in 2.7?

**Comment by jksuom** [comment]
> Does _sympify(True) not return S.true in 2.7?

It does in my system.
```
In [8]: srepr(_sympify(True))
Out[8]: 'S.true'
```
Must be something else...

**Comment by jksuom** [comment]
There was no problem:
```
sympy pr-10501$ bin/test test_relational
============================= test process starts ==============================
executable:         /usr/bin/python  (2.7.3-final-0) [CPython]
architecture:       64-bit
cache:              yes
ground types:       gmpy 1.14
random seed:        69899449
hash randomization: on (PYTHONHASHSEED=1271816623)

sympy/core/tests/test_relational.py[37] .....fff............................. 
                                                                            [OK]

======== tests finished: 34 passed, 3 expected to fail, in 5.56 seconds ========
```
I guess you should just restart the job.

**Comment by smichr** [comment]
Thanks for checking that out. I restarted. Let's see.

**Comment by smichr** [comment]
Thanks to @pbrady and the `rerun` option for tests, I was able to quickly identify that it was a Float problem that was causing an equality to sometimes fail. I think I have that fixed now by converting Float to Rational.

File: sympy/core/relational.py [hunk 1]
```diff
@@ -171,24 +171,171 @@ def equals(self, other, failing_expression=False):
...
+        forward = True
+        L, R = self.args
+        if not L:
+            L, R = R, L
+            forward = not forward
+
+        # simplify as long as it does not remmove any denominators
+        denL = [d for d in denoms(L) if not d.is_number]
+        _L = L.simplify(ratio=ratio, measure=measure)
+        _denL = [d for d in denoms(L) if not d.is_number]
```
Comment (jksuom) [inline sympy/core/relational.py]: Perhaps this should contain ``denoms(_L)``.

Comment (smichr) [inline sympy/core/relational.py]: got it (and on _denR, too)

**Comment by smichr** [comment]
This failing because after simplification, `factorial(x)/gamma(x)` is 1 and that looks like a denominator was lost. See note in test_simplify that was added with the modified test.

**Comment by smichr** [comment]
@jksuom, do you have an idea about how to handle this latest issue? I am trying to do "safe simplification" by not allowing simplify to return an expression which has lost a singularity defined by some denominator, e.g. `1/(1/x+1)` is not allowed to simplify to `x/(1 + x)`. But then this disallows `factorial(x)/gamma(x)` from simplifying to 1. I wonder if I should only disallow denominators which are **Add** instances to not disappear.

@jksuom wrote:

> 1/(1/x+1) is not allowed to simplify to x/(1 + x).

I'm afraid that trying to achieve this could result in problems hard to handle. In fact, there are reasons why I would rather allow this simplification.

BTW, you probably meant that ``factorial(x)/gamma(x)`` should simplify to ``x`` (for ``x`` a positive integer).

**Comment by smichr** [comment]
> you probably meant that factorial(x)/gamma(x) should simplify to x

yes (the test was `Eq(trig identity, factorial(x)/gamma(x))` which simplifies to `Eq(x, 1)`).

**Comment by smichr** [comment]
> there are reasons why I would rather allow this [`1/(1/x + 1)` -> `x/(x + 1)`] simplification

I would be interested to know what they are. The reason I am thinking to disallow this is that a Relational must be more careful than an expression that is being solved. In the latter case, one can use simplifications to find a solution and then *check* the solutions to eliminate any spurious roots. In the example I gave, the root of `x=0` would be rejected because it is not in the domain of the original function (please forgive mathematical inaccuracies, here). 

I was taking my cues from the you in this regard, btw.

If we do allow such simplifications then if we have `a = (1/x - x)*x` and `b=1 = x**2` (which is just `a.simplify()`) then `Eq(a - b, 0).simplify()` would be True -- and I see that it is in the PR(!) -- but for x=0 the lhs gives nan and the equality is False. And it is that which I am trying to avoid: giving a simplified result that would give a different answer than the non-simplified result.

Alternatively, we could return an **And** instance where condition under which the simplification is True are enumerated, e.g. using the `a` and `b` from the previous paragraph, `Eq(a - b, 0).simplify() -> And(True, Ne(x, 0)) -> Ne(x, 0)`.

**Comment by jksuom** [comment]
This is something that I haven't figured out in detail. It's more like a feeling, but I'll try to collect my thoughts.

``a = (1/x - x)*x`` and ``b = 1 - x**2`` are two different expressions, but I think they should be considered as denoting the same (rational) function. It is true that ``a`` gives ``nan`` at ``x = 0`` but that could be interpreted as "I was not able to compute the value", not as "there is no appropriate value". I think the evaluation of ``a`` should be improved to recognize 0 as a removable singularity and return the correct limit value 1. (This would be analogous to the way e.g. boolean expressions and integrals are often treated: if one method returns None, then try another one.)

In brief: I think that it would be preferable to improve the evaluation of expressions. There appear to be no clear grounds for making distinction between expressions denoting essentially same (abstract) objects.

**Comment by smichr** [comment]
@jksuom, are you expressing a different feeling now relative to the one that you expressed originally in your comment at #10401 ?  (Sorry, I don't know how to grab the link to send you directly there, but there is only one comment on the issue :-))

**Comment by jksuom** [comment]
That was about ``x/(x + 1)``. In fact, I was also then mentally identifying it with ``1/(1 + 1/x)``, and since the latter takes the value 1, I was feeling uneasy about the former. It is true that the value 1 is attained only at ``zoo``, but that seems to be generally not excluded (unless ``x = Symbol('x', finite=True)``).

**Comment by smichr** [comment]
But you said "of course, f != g" (where `f = x/(x + 1)` and `g = 1/(1 + 1/x)`. That they aren't the same is why I think we can't let the `Eq(f, g)` simplify to True; it is also why we don't report 0 as a solution to the one but do for the other:

```python
>>> var('x');g=1/(1/x+1)
x
>>> solve(g)
[]
>>> solve(g.simplify())
[0]
```

**Comment by jksuom** [comment]
> of course, f != g

I understood it was saying that f and g are *structurally* different. Anyway, that was the result when I tested it :-)

It is true that simplifying ``Eq(f, g)`` to True would currently lead to conflicts. But I was hoping that ``solve(g)`` could eventually be modified to yield ``[0]`` as ``g`` denotes a rational function.

**Comment by smichr** [comment]
@asmeurer , I remember some time ago discussing whether 0 is a solution to something like `1/(1/x + 1)` but have only found the references that I list in the OP. Do you recall anything else about the discussion or where it took place? I thought it had to do with whether or not a removable singularity was to be considered a solution to an equation.

**Comment by smichr** [comment]
OK, allowing removable singularities to disappear will simplify things, I think. In addition, I think I have a way to handle the case where a ratio of functions (like factorial and gamma) need to simplify. I'll try it some time this week.

**Pull request closed without merging**